sky/serve/infra_providers.py

```diff
@@ -1,222 +1,420 @@
+"""InfraProvider: handles the creation and deletion of endpoint replicas."""
+import base64
+import collections
+from concurrent import futures
 import logging
-from typing import List
-import time
+import multiprocessing
+import os
 import pickle
-import base64
+import requests
+import signal
+import threading
+import time
+from typing import List, Dict, Set, Optional, Any
 
 import sky
+from sky import backends
+from sky import status_lib
 from sky.backends import backend_utils
 
-import urllib
-import threading
-
 logger = logging.getLogger(__name__)
 
+_PROCESS_POOL_REFRESH_INTERVAL = 20
+_ENDPOINT_PROBE_INTERVAL = 10
+# TODO(tian): Maybe let user determine this threshold
+_REPLICA_UNHEALTHY_THRESHOLD_COUNTER = 180 // _ENDPOINT_PROBE_INTERVAL
+
 
 class InfraProvider:
+    """Each infra provider manages one services."""
+
+    def __init__(self,
+                 readiness_path: str,
+                 readiness_timeout: int,
+                 post_data: Optional[Any] = None) -> None:
+        self.ready_replicas: Set[str] = set()
+        self.unhealthy_replicas: Set[str] = set()
+        self.failed_replicas: Set[str] = set()
+        self.first_unhealthy_time: Dict[str, float] = dict()
+        self.continuous_unhealthy_counter: Dict[
```

Comment:
k8s term is `failureThreshold`. we can consider a similar term

Reply:
Renamed to `continuous_failure_counter` now. If you have any other ideas please let me know